### ISSUE TITLE
JACK/LV2 port metadata (for pretty names and CV ports)?

### ISSUE THREAD

**Issue by mxmilkiib** [grame-cncm/faust#435]
* https://jackaudio.org/metadata
* https://jackaudio.org/metadata/pretty-name
* https://jackaudio.org/metadata/signal-type
* https://github.com/jackaudio/jack2/blob/master/common/jack/metadata.h

The main purpose of signal-type is to hint/mark JACK ports as [CV](https://linuxmusicians.com/viewtopic.php?f=1&t=20701).

(N.b. there is also this request; https://bitbucket.org/agraef/faust-lv2/issues/22/support-for-lv2-cvport)

**Comment by mxmilkb** [comment]
@sletz Might you have any comment?

**Comment by sletz** [comment]
What is the question ?

**Comment by mxmilkb** [comment]
Feature request: it would be nice if, for JACK app* targets, the user could define information for setting JACK client metadata about the input/output port names and/or mark some as CV signal ports?

I guess LV2 targets could equally use this information for port names names and type (lv2:CVPort).

**Comment by sletz** [comment]
I'm not working on JACK anymore...

**Comment by mxmilkb** [comment]
For both JACK and LV2. I thought any LV2 effort works be done in the other project. Do any other plugin formats allow setting proper names for ports? Apologies, maybe pretty names and CV would be best approached as separate issues. In JACK, they are both just metadata properties, but maybe not just JACK is to be considered.

**Comment by sletz** [comment]
Be more concrete if possible: how to how suggest doing that?

**Comment by mxmilkb** [comment]
I would guess maybe using `declare` in the Standard Metadata. I'm not confident on what the notation format would be best for such a mapping though.

**Comment by sletz** [comment]
Please suggest something ((-;  What kind of string JACK meta-data system would be interested to get?

**Comment by mxmilkb** [comment]
https://jackaudio.org/api/group__Metadata.html has the exact details for JACK. The metadata link above covers using [jack_property](https://manpages.debian.org/testing/jackd1/jack_property.1.en.html) to make metadata changes.

(For LV2 it would be setting the port name and type Turtle metadata.)

**Comment by sletz** [comment]
So in the Faust DSP code:

declare ??
declare ??
declare  ??

**Comment by mxmilkb** [comment]
Essentially.

Pd-fork [https://github.com/Spaghettis/Spaghettis/pull/229](Spaghettis) uses messages.

Note: [Carla](https://github.com/falkTX/Carla), when hosting an LV2 bundle as a JACK client, goes further and presents that bundles port metadata as JACK port metadata, including the lv2:minimum and lv2:maximum, which is why Spaghettis mentions minimum and maximum.

(There is a the thought that, if CV signal is a normalised signal between -1 and 1, that the actual range it represents can be communicated out-of-band for some next level opportunities. There are no implementations/uses of this yet, and [MOD devices CV plugins](https://github.com/moddevices/mod-cv-plugins) uses a CV range that goes beyond -1 to 1)

**Comment by mxmilkb** [comment]
Though that does kind of depend on ones definition of "essentially"; maybe there is a better way of notating such metadata?

Might it be possible to have this reopened?

**Comment by mxmilkiib** [comment]
Can I request again that this issue be reopened?

Currently faust2lv2 outputs plugins with ports like:
![image](https://user-images.githubusercontent.com/108225/85930311-340ac800-b8b3-11ea-9ff4-6ea378286b12.png)

And faust2jack outputs apps with ports like:
![image](https://user-images.githubusercontent.com/108225/85930352-86e47f80-b8b3-11ea-9226-ce6f79dc5131.png)

But LV2 plugins can have ports like:
[![image](https://user-images.githubusercontent.com/108225/85930370-b85d4b00-b8b3-11ea-850f-e46198c8e19e.png)](https://user-images.githubusercontent.com/108225/85911333-91076f00-b81c-11ea-9b49-d0bae859a768.png)

And JACK apps like:
![image](https://user-images.githubusercontent.com/108225/85930374-c612d080-b8b3-11ea-9310-47d31687fb2e.png)

Nice looking port names is probably more a priority for general users, but I've a big interest in modular synths and recently watched the MOOC videos by @orlarey and @rmichon so I would quite like to try making 'CV' plugins in FAUST if it might be possible going forward?

**Comment by sletz** [comment]
Yes, this discussion about associating proper "semantic" to audio ins/outs is also quite important in the context of VCV Rack. We are indeed working on that, see: https://github.com/grame-cncm/faust/tree/vcvrack/architecture/vcvrack.

**Comment by cbix** [comment]
Now with the [vcvrack target](https://faustdoc.grame.fr/workshops/2020-11-21-faust-vcvrack/) released, the above questions seem mostly resolved and no changes to the language or existing practices are necessary:
> `[CV:N]` can be used in input (typically _sliders_ or _nentry_) or output (typically _bargraph_) controllers to connect them to CV instead of regular GUI parameters.

Now while with `faust2vcvrack`, input/output channels are automatically used as CV (there is no audio port type in Rack, just an [Audio module](https://library.vcvrack.com/Core/AudioInterface) interfacing with the outside world), this is not practical for JACK/LV2. The solution should only convert UI primitives tagged with `[CV:n]` to CV type ports and keep in/out channels as audio ports.

Ideally, any Faust code that compiles to vcvrack would then also compile to JACK and LV2, providing the same functionality.

**Comment by SpotlightKid** [comment]
@cbix: I'm not sure how your remarks relate to (general) metadata for input/output ports. Can you elaborate?

**Comment by cbix** [comment]
@SpotlightKid not related to general metadata. However one of the requests was to allow CV ports, which JACK supports by metadata and LV2 supports as a port type. Now that there is a solution for CV ports in another target (vcvrack), this could be used _instead_ of generic port metadata to enable CV.

For all other use cases of generic channel metadata outside CV I'd still support other solutions.
